services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowSyncSpec.groovy

```diff
@@ -0,0 +1,188 @@
+
+    @Autowired
+    TopologyDefinition topology
+    @Autowired
+    FlowHelper flowHelper
+    @Autowired
+    NorthboundService northboundService
+    @Autowired
+    PathHelper pathHelper
+    @Autowired
+    IslUtils islUtils
+    @Autowired
+    Database db
+
+    @Shared
+    int flowRulesCount = 2
+
+    def "Able to synchronize a flow (install missing flow rules, reinstall existing) without rerouting"() {
+        given: "An intermediate-switch flow with one deleted rule on each switch"
+        def switches = topology.getActiveSwitches()
+        def allLinks = northboundService.getAllLinks()
+        def (Switch srcSwitch, Switch dstSwitch) = [switches, switches].combinations()
+                .findAll { src, dst -> src != dst }.find { Switch src, Switch dst ->
+            allLinks.every { link ->
+                def switchIds = link.path*.switchId
+                !(switchIds.contains(src.dpId) && switchIds.contains(dst.dpId))
+            }
+        } ?: assumeTrue("No suiting switches found to build an intermediate-switch flow.", false)
+
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flowHelper.addFlow(flow)
+        def flowPath = PathHelper.convert(northboundService.getFlowPath(flow.id))
+
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id)
+        def ruleToDelete = getFlowRules(involvedSwitches.first().dpId).first().cookie
+        involvedSwitches.each { sw ->
+            northboundService.deleteSwitchRules(sw.dpId, ruleToDelete)
+            Wrappers.wait(RULES_DELETION_TIME) {
+                assert getFlowRules(sw.dpId).size() == flowRulesCount - 1
+            }
+        }
+
+        when: "Synchronize the flow"
+        Map<SwitchId, Long> rulesDurationMap = involvedSwitches.collectEntries {
+            [(it.dpId): getFlowRules(it.dpId).first().durationSeconds]
+        }
+
+        def syncTime = new Date()
+        def rerouteResponse = northboundService.synchronizeFlow(flow.id)
+
+        then: "The flow is not rerouted"
+        int seqId = 0
+
+        !rerouteResponse.rerouted
+        rerouteResponse.path.path == flowPath
+        rerouteResponse.path.path.each { assert it.seqId == seqId++ }
+
+        PathHelper.convert(northboundService.getFlowPath(flow.id)) == flowPath
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.UP }
+
+        and: "Missing flow rules are installed (existing ones are reinstalled) on all switches"
+        involvedSwitches.each { sw ->
+            def flowRules = []
+            Wrappers.wait(RULES_INSTALLATION_TIME) {
+                flowRules = getFlowRules(sw.dpId)
+                assert flowRules.size() == flowRulesCount
+                flowRules.each {
+                    assert it.durationSeconds < rulesDurationMap[sw.dpId] +
+                            TimeCategory.minus(new Date(), syncTime).toString().replace(" seconds", "").toFloat()
```

Comment:
Do `TimeCategory.minus(new Date(), syncTime).seconds` instead

Reply:
If you care about precision here, then sth like `TimeCategory.minus(new Date(), syncTime).millis / 1000.0` should do the trick. Working with strings looks too dirty here

Reply:
Thanks, fixed.